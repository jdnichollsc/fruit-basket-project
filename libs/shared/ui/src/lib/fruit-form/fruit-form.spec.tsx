import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { FruitForm } from '.';
import type { FruitFormProps } from '.';

describe('FruitForm', () => {
  const mockProps: FruitFormProps = {
    onAdd: jest.fn(),
    isLoading: false,
    error: null
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render form with accessible elements', () => {
    render(<FruitForm {...mockProps} />);
    
    expect(screen.getByRole('textbox', { name: /fruit name/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /add/i })).toBeInTheDocument();
  });

  it('should use default isLoading value when not provided', () => {
    const { onAdd, error } = mockProps;
    render(<FruitForm onAdd={onAdd} error={error} />);
    
    expect(screen.getByRole('textbox', { name: /fruit name/i })).not.toBeDisabled();
    expect(screen.getByRole('button', { name: /add/i })).not.toBeDisabled();
  });

  it('should disable form elements when loading', () => {
    render(<FruitForm {...mockProps} isLoading={true} />);
    
    expect(screen.getByRole('textbox', { name: /fruit name/i })).toBeDisabled();
    expect(screen.getByRole('button', { name: /add/i })).toBeDisabled();
  });

  it('should handle fruit submission', async () => {
    const user = userEvent.setup();
    render(<FruitForm {...mockProps} />);
    
    const input = screen.getByRole('textbox', { name: /fruit name/i });
    const button = screen.getByRole('button', { name: /add/i });
    
    await user.type(input, 'Mango');
    await user.click(button);
    
    expect(mockProps.onAdd).toHaveBeenCalledWith('Mango');
    expect(input).toHaveValue('');
  });

  it('should display error message', () => {
    const error = 'Fruit already exists';
    render(<FruitForm {...mockProps} error={error} />);
    
    expect(screen.getByRole('alert')).toHaveTextContent(error);
  });

  it('should not submit empty or whitespace-only values', async () => {
    const user = userEvent.setup();
    render(<FruitForm {...mockProps} />);
    
    const input = screen.getByRole('textbox', { name: /fruit name/i });
    const button = screen.getByRole('button', { name: /add/i });
    
    // Empty submission
    await user.click(button);
    expect(mockProps.onAdd).not.toHaveBeenCalled();
    
    // Whitespace-only submission
    await user.type(input, '   ');
    await user.click(button);
    expect(mockProps.onAdd).not.toHaveBeenCalled();
  });

  it('should handle form submission via Enter key', async () => {
    const user = userEvent.setup();
    render(<FruitForm {...mockProps} />);
    
    const input = screen.getByRole('textbox', { name: /fruit name/i });
    
    await user.type(input, 'Mango{Enter}');
    
    expect(mockProps.onAdd).toHaveBeenCalledWith('Mango');
    expect(input).toHaveValue('');
  });
}); 