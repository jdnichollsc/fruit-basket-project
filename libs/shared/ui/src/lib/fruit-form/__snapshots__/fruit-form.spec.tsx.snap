// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FruitForm should match snapshot 1`] = `
<div>
  <form
    class="mb-6"
  >
    <div
      class="flex gap-2"
    >
      <label
        class="flex-1 form-control"
      >
        <span
          class="sr-only"
        >
          Fruit name
        </span>
        <input
          aria-label="Fruit name"
          class="input input-bordered w-full"
          data-testid="add-fruit-input"
          placeholder="Enter fruit name"
          required=""
          type="text"
          value=""
        />
      </label>
      <button
        class="btn btn-primary min-w-20"
        type="submit"
      >
        Add
      </button>
    </div>
  </form>
</div>
`;

exports[`FruitForm should match snapshot when loading 1`] = `
<div>
  <form
    class="mb-6"
  >
    <div
      class="flex gap-2"
    >
      <label
        class="flex-1 form-control"
      >
        <span
          class="sr-only"
        >
          Fruit name
        </span>
        <input
          aria-label="Fruit name"
          class="input input-bordered w-full input-disabled"
          data-testid="add-fruit-input"
          disabled=""
          placeholder="Enter fruit name"
          required=""
          type="text"
          value=""
        />
      </label>
      <button
        class="btn btn-primary min-w-20 btn-disabled"
        disabled=""
        type="submit"
      >
        <span
          class="loading loading-spinner"
        />
        Adding...
      </button>
    </div>
  </form>
</div>
`;

exports[`FruitForm should match snapshot with error 1`] = `
<div>
  <form
    class="mb-6"
  >
    <div
      class="flex gap-2"
    >
      <label
        class="flex-1 form-control"
      >
        <span
          class="sr-only"
        >
          Fruit name
        </span>
        <input
          aria-label="Fruit name"
          class="input input-bordered w-full"
          data-testid="add-fruit-input"
          placeholder="Enter fruit name"
          required=""
          type="text"
          value=""
        />
      </label>
      <button
        class="btn btn-primary min-w-20"
        type="submit"
      >
        Add
      </button>
    </div>
    <p
      class="mt-2 text-sm text-error"
      role="alert"
    >
      Test error message
    </p>
  </form>
</div>
`;
